File: EarthquakeList/src/main/java/com/jseo/earthquakelist/DataRetriever.java
package com.jseo.earthquakelist;

public abstract class DataRetriever {

    private DataParser mParser;
    private OnRetrieveCompleteListener mOnCompleteListener;
    private String mUrlString;


    public interface OnRetrieveCompleteListener {
        void onRetrieveComplete(boolean isSuccess, Object retrievedData);
    }

    public DataRetriever() {
        mParser = null;
        mOnCompleteListener = null;
        mUrlString = null;
    }

    public DataRetriever(OnRetrieveCompleteListener listener) {
        mParser = null;
        mOnCompleteListener = listener;
        mUrlString = null;
    }

    public void setOnRetrieveCompleteListener(OnRetrieveCompleteListener listener) {
        mOnCompleteListener = listener;
    }

    public OnRetrieveCompleteListener getOnRetrieveCompleteListener() {
        return mOnCompleteListener;
    }

    public void setUrlString(String urlString) {
        mUrlString = urlString;
    }

    public String getUrlString() {
        return mUrlString;
    }

    public void setDataParser(Object data) {
        mParser = DataParserFactory.createDataParser(data);
        mParser.setOnParseCompleteListener(new DataParser.OnParseCompleteListener() {
            @Override
            public void onParseComplete(boolean isSuccess, Object parsedData) {
                if (mOnCompleteListener != null) {
                    mOnCompleteListener.onRetrieveComplete(isSuccess, parsedData);
                }
            }
        });
    }

    /**
     * derived Retriever should implement how to retrieve data
     */
    public abstract void retrieve();

    void parseData() throws Exception {
        if (mParser != null) {
            mParser.parse();
        } else {
            throw new Exception("Parser is not set!");
        }
    }
}

File: EarthquakeList/src/main/java/com/jsbomb/earthquakelist/actors/DataRetriever.kt
package com.jsbomb.earthquakelist.actors

import kotlinx.coroutines.CoroutineScope

abstract class DataRetriever(
    private var parser: DataParser? = null,
    private var onCompleteListener: OnRetrieveCompleteListener? = null,
    private var urlString: String? = null
) {

    interface OnRetrieveCompleteListener {
        /**
         * callback to be called after retrieval (and parsing) is complete. make sure to update UI
         * in the main thread
         */
        fun onRetrieveComplete(isSuccess: Boolean, retrievedData: Any?)
    }

    fun setOnRetrieveCompleteListener(listener: OnRetrieveCompleteListener) {
        onCompleteListener = listener
    }

    fun getOnRetrieveCompleteListener(): OnRetrieveCompleteListener? = onCompleteListener

    fun setUrlString(urlString: String) {
        this.urlString = urlString
    }

    fun getUrlString(): String? = urlString

    fun setDataParser(parser: DataParser) {
        this.parser = parser
    }

    fun attachData(data: Any) {
        val dataParser = DataParserFactory.createDataParser(data).apply {
            setOnParseCompleteListener(object : DataParser.OnParseCompleteListener {
                override fun onParseComplete(isSuccess: Boolean, parsedData: Any?) {
                    callOnRetrieveComplete(isSuccess, parsedData)
                }
            })
        }
        setDataParser(dataParser)
    }

    protected fun callOnRetrieveComplete(isSuccess: Boolean, parsedData: Any?) {
        onCompleteListener?.onRetrieveComplete(isSuccess, parsedData)
    }

    /**
     * derived Retriever should implement how to retrieve data
     */
    abstract suspend fun retrieve(): Any?

    fun parseRetrievedData() {
        parser?.parse()
    }
}